updater.go

```diff
@@ -126,23 +130,31 @@ func (u *Updater) update(ctx Context, options UpdateOptions) (*Update, *Error) {
 		return update, verifyErrPtr(err)
 	}
 
-	if err := ctx.BeforeApply(*update); err != nil {
-		return update, applyErrPtr(err)
+	if err := u.apply(ctx, *update, options); err != nil {
+		return update, err
 	}
 
-	if err := u.platformApplyUpdate(*update, options); err != nil {
-		return update, applyErrPtr(err)
+	if err := ctx.Restart(); err != nil {
+		return update, restartErrPtr(err)
```

Comment:
100% in support of type switch or type assertion checking of errors. It's definitely the golang way, and it's the standard way of handling not just errors but any time you need to resolve an interface.